pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,390 @@
+        top_k: number of highest probability entries for each sample to convert to 1s, relevant
+            only for (multi-dimensional) multi-class cases.
+        is_multiclass: if True, treat binary and multi-label inputs as multi-class or multi-dim
+            multi-class with 2 classes, respectively. If False, treat multi-class and multi-dim
+            multi-class inputs with 1 or 2 classes as binary and multi-label, respectively.
+            Defaults to None, which treats inputs as they appear.
+
+    Returns:
+        preds: binary tensor of shape (N, C) or (N, C, X)
+        target: binary tensor of shape (N, C) or (N, C, X)
+    """
+    preds, target = preds.clone().detach(), target.clone().detach()
+
+    # Remove excess dimensions
+    if preds.shape[0] == 1:
+        preds, target = preds.squeeze().unsqueeze(0), target.squeeze().unsqueeze(0)
+    else:
+        preds, target = preds.squeeze(), target.squeeze()
+
+    _check_classification_inputs(
+        preds,
+        target,
+        threshold=threshold,
+        num_classes=num_classes,
+        is_multiclass=is_multiclass,
+        top_k=top_k,
+    )
+
+    preds_float = preds.is_floating_point()
+
+    if preds.ndim == target.ndim == 1 and preds_float:
+        mode = "binary"
+        preds = (preds >= threshold).int()
+
+        if is_multiclass:
+            target = to_onehot(target, 2)
+            preds = to_onehot(preds, 2)
+        else:
+            preds = preds.unsqueeze(-1)
+            target = target.unsqueeze(-1)
+
+    elif preds.ndim == target.ndim and preds_float:
+        mode = "multi-label"
+        preds = (preds >= threshold).int()
+
+        if is_multiclass:
+            preds = to_onehot(preds, 2).reshape(preds.shape[0], 2, -1)
+            target = to_onehot(target, 2).reshape(target.shape[0], 2, -1)
+        else:
+            preds = preds.reshape(preds.shape[0], -1)
+            target = target.reshape(target.shape[0], -1)
+
+    elif preds.ndim == target.ndim + 1 == 2:
+        mode = "multi-class"
+        if not num_classes:
+            num_classes = preds.shape[1]
+
+        target = to_onehot(target, num_classes)
+        preds = select_topk(preds, top_k)
+
+        # If is_multiclass=False, force to binary
+        if is_multiclass is False:
+            target = target[:, [1]]
+            preds = preds[:, [1]]
+
+    elif preds.ndim == target.ndim == 1 and not preds_float:
+        mode = "multi-class"
+
+        if not num_classes:
+            num_classes = max(preds.max(), target.max()) + 1
+
+        # If is_multiclass=False, force to binary
+        if is_multiclass is False:
+            preds = preds.unsqueeze(1)
+            target = target.unsqueeze(1)
+        else:
+            preds = to_onehot(preds, num_classes)
+            target = to_onehot(target, num_classes)
+
+    # Multi-dim multi-class (N, ...) with integers
+    elif preds.shape == target.shape and not preds_float:
+        mode = "multi-dim multi-class"
```

Comment:
just an idea, shall we use , as a separator?
```suggestion
        mode = "multi-dim,multi-class"
```

Reply:
I like it better the way it is - corresponds more to the fully written out word (multi-dimensional multi-class)

Reply:
is it `multi-class (N, C, ...)`??

Reply:
In this case it is `(N, ...)`

Reply:
ok, it's for labels. :+1:

Reply:
I agree with Borda, should be separable by a separator.

Reply:
nothing bad, just a suggestion :]